### PR TITLE
PETSc: split install() into configure(), build(), install()

### PR DESCRIPTION
**Pull request by JDBetteridge**
PETSc has its [own build system](https://petsc.org/main/install/), but it fits a `./configure`, `make`, `make install` pattern.

Currently all the logic for a PETSc install is included in the `install()` method of the package. If another package inherits from this all of the logic must be reproduced.

This PR splits the PETSc installation into minimal `configure()`, `build()` and `install()` methods.

All the logic determining the correct arguments for the configure step is moved to the `configure_options()` so it can be reproduced easily in a subclass.

I don't fully understand why the package reverts changes by kokkos-nvcc-wrapper, and I'm not sufficiently familiar with the interplay between PETSc, Kokkos and Cuda. The method `revert_kokkos_nvcc_wrapper()` just reproduces the behaviour of the original package in the configure, build and install stages.

### CODE REVIEW

**Comment by JDBetteridge** [comment]
@balay @BarrySmith @jedbrown any chance of a review on this?
I was hoping it wasn't too controversial!

Many thanks

**Review by jedbrown — DISMISSED** [review]
Looks okay to me. @balay or @jczhang07 should be able to explain the Kokkos issue.

**Comment by balay** [comment]
Perhaps you can use `spec = self.spec` at the beginning of these routines so that the diff is minimal [and the change is clearer]

**Comment by balay** [comment]
> I don't fully understand why the package reverts changes by kokkos-nvcc-wrapper

`kokkos-nvcc-wrapper` is changing the mpicc/mpicxx etc compilers [to use this wrapper] - and this breaks PETSc builds - ence this is reverted here in petsc spec.

And I don't remember the exact errors. I see it was added in https://github.com/spack/spack/pull/27624 - so presumably should be reproducible with `spack install petsc+cuda+kokkos` - with that change removed from petsc spec.

Related `kokkos-nvcc-wrapper` issues [with rocm] were at https://github.com/spack/spack/issues/27684

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 1]
```diff
@@ -403,14 +406,17 @@ def install(self, spec, prefix):
...
         else:
             options.append('--with-clanguage=C')
 
         # Activates library support if needed (i.e. direct dependency)
         jpeg_sp = spec['jpeg'].name if 'jpeg' in spec else 'jpeg'
-        scalapack_sp = spec['scalapack'].name if 'scalapack' in spec else 'scalapack'
+        if 'scalapack' in spec:
+            scalapack_sp = spec['scalapack'].name
+        else:
+            scalapack_sp = 'scalapack'
```
Comment (balay) [inline var/spack/repos/builtin/packages/petsc/package.py]: I suspect this change can also be reverted

Comment (JDBetteridge) [reply var/spack/repos/builtin/packages/petsc/package.py]: Yes, it just sneaks inside the flake8 line length!

**Review by balay — APPROVED** [review]

**Comment by balay** [comment]
@spackbot run pipeline

**Comment by balay** [comment]
@spackbot run pipeline

**Pull request merged**